R/toolGetMapping.R

```diff
@@ -24,58 +24,87 @@
+    fname <- .searchNameLocal(name, type = type)
+    if (file.exists(as.character(fname))) {
+      return(fname)
+    }
+    fname <- .searchNameMappingFolder(name, type = type)
+    if (file.exists(as.character(fname))) {
+      return(fname)
+    }
+
+    packages <- getConfig("packages")
+    if (!is.null(activecalc[[1]]) && any(grepl(paste0("^", activecalc[[1]], "$"), getCalculations()[, "type"]))) {
+      fp <- as.character(attr(prepFunctionName(activecalc[[1]], "calc"), "package"))
+      packages <- c(fp, grep(fp, packages, invert = TRUE, value = TRUE))
+    }
+    for (i in packages) {
+      out <- .searchNamePackage(name, type = type, packageName = i)
+      if (out != "") {
+        fname <- out
       }
     }
+    # if not found in packages, fall back to result from mappingfolder, even though it doesn't exist
+    return(fname)
+
   } else if (where == "mappingfolder") {
-    mf <- getConfig("mappingfolder")
-    fname <- paste0(mf, "/", type, "/", name)
+    return(.searchNameMappingFolder(name = name, type = type))
   } else if (where == "local") {
-    if (is.null(type)) {
-      fname <- name
-    } else {
-      fname <- paste0(type, "/", name)
-    }
+    return(.searchNameLocal(name = name, type = type))
   } else {
-    if (is.null(type)) {
-      tmpfname <- name
-    } else {
-      tmpfname <- paste0(type, "/", name)
-    }
-    fname <-  system.file("extdata", tmpfname, package = where)
-    if (fname == "") fname <- system.file("inst/extdata", tmpfname, package = where)
-    if (fname == "") fname <- system.file("extdata", strsplit(tmpfname, split = "/")[[1]][2], package = where)
-    if (fname == "") fname <- system.file("inst/extdata", strsplit(tmpfname, split = "/")[[1]][2], package = where)
-    if (fname == "" & error.missing) {
-      stop('Mapping "', name, '" with type "', type, '" not found in package "', where, '"!')
-    }
+    return(.searchNamePackage(name = name, type = type, packageName = where))
   }
-  if (error.missing & !file.exists(as.character(fname))) stop('Mapping "', name, '" not found!')
-  fname <- gsub("/+", "/", fname)
-  if (returnPathOnly) return(fname)
+}
+
+.searchNameLocal <- function(name, type) {
+  if (file.exists(as.character(name))) {
+    return(name)
+  }
+  return(.typedName(name, type))
+}
+
+.searchNameMappingFolder <- function(name, type) {
+  mf <- getConfig("mappingfolder")
+  return(paste0(mf, "/", type, "/", name))
+}
+
+.searchNamePackage <- function(name, type, packageName) {
+  tmpfname <- .typedName(name, type)
+  fname <- system.file("extdata", tmpfname, package = packageName)
+  if (fname == "" && !is_dev_package(packageName)) fname <- system.file("inst/extdata", tmpfname, package = packageName)
+  if (fname == "") fname <- system.file("extdata", strsplit(tmpfname, split = "/")[[1]][2], package = packageName)
+  if (fname == "" && !is_dev_package(packageName))
+    fname <- system.file("inst/extdata", strsplit(tmpfname, split = "/")[[1]][2], package = packageName)
```

Comment:
Here are the new checks for `is_dev_package` - if it is, `system.file("inst/extdata")` throws an error because stuff under `inst/` is installed at the package top level, so this `system.file` call should never find anything. To be perfectly bug-compatible with the current madrat in production, I introduced the `!is_dev_package` guard. But *most likely* the `system.file("inst/…")` calls were introduced in error and they could be removed instead. That would make things less weird. What do you think?

Reply:
Interestingly, the `system.file("inst/…")` call(s) were introduced in https://github.com/pik-piam/madrat/commit/6f9d872425bc6aacfaba1a2c2947d82a033b5f60 due to a "bug in `system.file`", so apparently 5 years ago it was necessary to ask for "inst/…"? But when was `system.file` fixed? Can we drop this workaround or do we need to pile the other workaround onto this one?